src/Microsoft.AspNetCore.Authentication2.Cookies/CookieAuthenticationHandler.cs

```diff
@@ -0,0 +1,461 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+
+using System;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.Authentication2.Cookies
+{
+    public class CookieAuthenticationHandler : AuthenticationSchemeHandler<CookieAuthenticationOptions>
+    {
+        private const string HeaderValueNoCache = "no-cache";
+        private const string HeaderValueMinusOne = "-1";
+        private const string SessionIdClaim = "Microsoft.AspNetCore.Authentication2.Cookies-SessionId";
+
+        private bool _finishCalled;
+        private bool _shouldRefresh;
+        private DateTimeOffset? _refreshIssuedUtc;
+        private DateTimeOffset? _refreshExpiresUtc;
+        private string _sessionKey;
+        private Task<AuthenticateResult> _readCookieTask;
+
+        public async override Task InitializeAsync(AuthenticationScheme scheme, HttpContext context)
+        {
+            await base.InitializeAsync(scheme, context);
+
+            // TODO: This needs to go into some kind of base class for reuse
+            if (Options.EventsType != null)
+            {
+                Options.Events = context.RequestServices.GetRequiredService(Options.EventsType) as CookieAuthenticationEvents;
+            }
+
+            if (Options.Events == null)
+            {
+                Options.Events = new CookieAuthenticationEvents();
+            }
+            if (String.IsNullOrEmpty(Options.CookieName))
+            {
+                Options.CookieName = CookieAuthenticationDefaults.CookiePrefix + Scheme.Name;
```

Comment:
I agree with @Tratcher. It's not only weird, it's dangerous (and potentially a perf' disaster if you receive thousands of requests when the app warms up, as there's no locking preventing `InitializeAsync` from being called multiple times).

There's also another BIG problem with this design approach: configuration errors cannot be detected at startup and won't prevent the app from loading successfully (so each time it will receive a new request, it will try to initialize the handlers... and fail miserably).